Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractCompactRandomCutTree.java

```diff
@@ -150,9 +103,80 @@ protected void addSequenceIndex(Integer nodeRef, long sequenceIndex) {
+                    int head = nodeQueue.poll();
+                    int newLeft;
+                    int leftChild = getLeftChild(head);
+                    if (isLeaf(leftChild)) {
+                        // the parent is the current node and the indices, mass are being copied over
+                        map[leftChild] = result.addLeaf(currentNode, getPointReference(leftChild), getMass(leftChild));
+                        validateInternalState(map[leftChild] == leafcounter, "incorrect state");
+                        newLeft = leafcounter++;
+                    } else { // leftchild is an internal node
+                        newLeft = ++nodeCounter;
+                        nodeQueue.add(leftChild);
+                    }
+                    int newRight;
+                    int rightChild = getRightChild(head);
+                    if (isLeaf(rightChild)) {
+                        // the parent is the current node and the indices, mass are being copied over
+                        map[rightChild] = result.addLeaf(currentNode, getPointReference(rightChild),
+                                getMass(rightChild));
+                        validateInternalState(map[rightChild] == leafcounter, "incorrect state");
+                        newRight = leafcounter++;
+                    } else {
+                        newRight = ++nodeCounter;
+                        nodeQueue.add(rightChild);
+                    }
+                    // the parent has to be added by now
+                    int parent = (head == root) ? -1 : map[getParent(head)];
+                    map[head] = result.addNode(parent, newLeft, newRight, getCutDimension(head), getCutValue(head),
+                            getMass(head));
+                    validateInternalState(map[head] == currentNode, "incorrect state");
+                    currentNode++;
+                }
+                validateInternalState(currentNode == nodeStore.size(), "incorrect state");
+
+                if (cachedBoxes != null) {
+                    swapCaches(map);
+                }
+                if (enableSequenceIndices) {
+                    HashMap<Long, Integer>[] newSequence = new HashMap[maxSize];
+                    for (int i = 0; i < maxSize; i++) { // iterate over leaves
+                        if (map[i + maxSize - 1] != NULL) { // leaf is in use
+                            validateInternalState(isLeaf(map[i + maxSize - 1]), "error in map");
+                            newSequence[map[i + maxSize - 1] - maxSize + 1] = sequenceIndexes[i];
+                        }
+                    }
+                    sequenceIndexes = newSequence;
+                }
+                root = 0;
+            } else {
+                root = result.addLeaf(NULL, getPointReference(root), getMass(root));
+            }
+        }
+        nodeStore = result;
+    }
+
     @Override
     protected void deleteSequenceIndex(Integer nodeRef, long sequenceIndex) {
```

Comment:
Should we move sequence indexes into the node store while we're at it? Then `computeLeafIndex` wouldn't have to be a public method.

Reply:
True, but we have to keep PointSum and the boxes in the tree (precision issues), and thus any reconfiguration has to be exposed to the tree anyways -- the NodeStore now just has information about internal nodes and leaves. It may be easier to get a different implementation of NodeStore if needed (if we keep the sequenceIndex separate, at least for now). This decision would not affect the state in any way.